File: SpiralArray.py
# Time Complexity : O(M x N) //As it would run for each row and column of the array
#  Space Complexity : O(1) //As we are using the same array to return the result
#  Did this code successfully run on Leetcode : Yes
#  Any problem you faced while coding this : N/A

#  Your code here along with comments explaining your approach
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        #Declare the bounds of array
        m = len(matrix) 
        n= len(matrix[0]) 
        
        #Declare our result array and the directions
        res = []
        t,l=0,0
        r,b = n-1,m-1
        
        #Iterate till the top corsses the botom or the left crosses the right
        while t <= b and l <= r:
            
            #top row
            for i in range(l,r+1):
                res.append(matrix[t][i])
            t+=1        #move the top position to row below
            
            #right col
            for i in range(t,b+1):
                res.append(matrix[i][r])
            r-=1        #move the right position to column to the left
            
            #bottom row
            if t <= b:
                for i in range(r,l-1,-1):
                    res.append(matrix[b][i])
                b-=1     #move the bottom position to row above
            
            #left row
            if l <= r:
                for i in range(b,t-1,-1):
                    res.append(matrix[i][l])
                l+=1    #move the left position to column to the right
        return res

File: src/SpiralArray.java
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes

class SpiralArray {
    public List<Integer> getSpiralArray(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        int left = 0;
         int right = n-1;
        int top = 0;
         int bottom = m-1;
        List<Integer> list = new ArrayList<>();
        while(left <= right && top <= bottom){
            //top
            for(int j=left; j<=right; j++){
                list.add(matrix[top][j]);
            }
            top++;
            if(left <= right && top <= bottom){
                //bottom
                for(int i=top; i<=bottom; i++){
                    list.add(matrix[i][right]);
                }
                right--;
            }
            //right
            if(left <= right && top <= bottom){
                for(int j=right; j>=left; j--){
                    list.add(matrix[bottom][j]);
                }
                bottom--;
            }
            //left
            if(left <= right && top <= bottom){
                for(int i=bottom; i>=top; i--){
                    list.add(matrix[i][left]);
                }
                left++;
            }

        }
        return list;
    }
}